### DISCUSSION TITLE
Accesibility and keyboard navigation

### DISCUSSION THREAD

**Discussion by justbennet** [squidfunk/mkdocs-material#3156] [Q&A]
I am extremely happy with MkDocs and with the Material theme myself.  I am trying to make my site as friendly and usable to the blind and visually impaired as possible.  I met yesterday with someone from our university's accessibility group to look at a draft of our website to evaluate accessibility, and there were a few issues.  The one that is of most concern to me now is that when the table of contents has subtopics, there doesn't seem to be a way to navigate to them via the keyboard; i.e., Tabbing to them.

To illustrate, make a site with just

`mkdocs.yml`
```
site_name: "Accessibility test"
theme:
  name: material
nav:
  - Help: text.md
  - Topics:
    - Subtopic: text.md
  - More help: text.md
```

`docs/text.md`
```
# Text page

This page contains some text.
```

Open the page in a browser, and try to Tab to the Subtopic.  If the Topics item is expanded -- that is, Subtopic is visible -- you can Tab to it, whereas if the Topic is folded up -- that is, Subtopic is not visible -- I can find no way to get to Subtopic using the keyboard.

Is there a way to do that?

We would like to make our documentation functional for screen reading software and keyboard navigation without sacrificing the item headings in the Navigation bar.

Thank you for your time and for the great software!

**Comment by squidfunk** [comment]
Thanks for reporting. A11Y is a topic of great importance for Material for MkDocs, but I agree that some parts of the UI could use some work. You can open an issue if you like where we can collect A11Y issues. I can look into them when I find some time.

Related:

- #1953
- #1665
- #1640
- #1514
- [and more...](https://github.com/squidfunk/mkdocs-material/issues?q=accessibility)

**Reply by justbennet** [reply]
Thanks.  I wasn't sure whether this was an issue or I just missed the 'trick' to doing this.  I will convert this to an issue.

There are a couple of other accessibility issues that were raised.

- Section headings have permalinks as links, which is likely to be confusing to screen readers, as it will be a link to itself but the screen reader is unlikely to be able to differentiate it from a link to somewhere else.  Not sure whether this is a 'problem' to be 'solved'.  They are all also labeled 'Permanent link'.  Would there be a way to relabel them so that they are maybe labeled something like 'Anchor link to this section for reference'?
- The page title changes to match the current section heading in the page body, which might also mislead a screen reader when a user is in section and tabs back to the top.

There was something about the use of 'titles', 'alt', and 'aria-???' for labeling images, the details of which I do not have to hand.

Some of those might need to be dealt with via the markdown renderer not the theme, but should each go into a separate issue if/when I can make small examples?

Also, would it make sense to have an omnibus discussion topic for accessibility from which specific issues could branch but might make it easier for people wanting to follow the general topic easier to locate relevant discussion?  Just a thought.

**Reply by squidfunk** [reply]
> Section headings have permalinks as links, which is likely to be confusing to screen readers, as it will be a link to itself but the screen reader is unlikely to be able to differentiate it from a link to somewhere else. Not sure whether this is a 'problem' to be 'solved'. They are all also labeled 'Permanent link'. Would there be a way to relabel them so that they are maybe labeled something like 'Anchor link to this section for reference'?

Yes, they can be relabeled, see `permalink_title` in the [Table of Contents docs](https://python-markdown.github.io/extensions/toc/#usage).

> The page title changes to match the current section heading in the page body, which might also mislead a screen reader when a user is in section and tabs back to the top.

This cannot be disabled. If we can add `aria-*` attributes to improve it, I'm happy to collaborate with you on a PR.

> There was something about the use of 'titles', 'alt', and 'aria-???' for labeling images, the details of which I do not have to hand.

Labeling images can be done by the author using the Attribute Lists extension, similar to image alignment, [see the docs](https://squidfunk.github.io/mkdocs-material/reference/images/#image-alignment).

> Some of those might need to be dealt with via the markdown renderer not the theme, but should each go into a separate issue if/when I can make small examples? Also, would it make sense to have an omnibus discussion topic for accessibility from which specific issues could branch but might make it easier for people wanting to follow the general topic easier to locate relevant discussion? Just a thought.

Please create a single issue that lists all A11Y problems that I haven't answered/solved with this comment. If applicable, use a checklist in your OP, so we can mark off all problems as we solve them. You can then add separate sections in your OP or comments with problem descriptions and reproducible cases. I know that this is not how the issue tracker was intended, but I don't want the issue tracker to be flooded with A11Y issues which I consider all to be enhancements (not bugs).

**Reply by squidfunk** [reply]
I've added a [short description](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#toc-permalink-title) for the `permalink_title` option in 8f382d979.

**Reply by justbennet** [reply]
Thanks for that.  Very helpful.

Expanding all the sections, as suggested in https://github.com/squidfunk/mkdocs-material/discussions/3161 is a kind of workaround, as then all the sections are then navigable via Tab.  That isn't a good solution for many, I think.

I speak HTML, CSS, and Javascript only with the help of a phrasebook, so pardon my ignorance.

I am looking at the HTML that is generated, and for a regular navigation bar  item, it looks like this.

```
    <li class="md-nav__item">
      <a href="xsede/" class="md-nav__link">
        XSEDE resources
      </a>
    </li>
 ```

and for a navigation with nested items, it looks like

 ```
        <li class="md-nav__item md-nav__item--nested">
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_7" type="checkbox" id="__nav_7" >
      <label class="md-nav__link" for="__nav_7">
        Using R
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using R" data-md-level="1">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          Using R
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
    <li class="md-nav__item">
      <a href="R/R_kernel/" class="md-nav__link">
        Installing the R kernel for Jupyter
      </a>
    </li>      
        </ul>
      </nav>
    </li>
```

I don't know enough about HTML to know how to turn the label and the `<span class="md-nav__icon md-icon"></span>` into a Tab-navigable item.  That seems to be a checkbox, and if it were a Tab-navigable item, then pressing Return would expand it, yes?  As it stands, I think it relies on Javascript focus and hover to be activated.

I'd like to help, but I may be too ignorant to offer much.

**Reply by squidfunk** [reply]
I think in the end it might be a combination of JavaScript and adding `tabIndex` attributes to elements.

**Comment by mnazzaro** [comment]
I wrote a fix here: https://github.com/mnazzaro/mkdocs-material

I added partials/javascripts/nav-item.html and updated partials/nav-item.html and base.html so that navigation labels would have accessibility tags and tabindex="0". Previously, they weren't tabbable.

I also  wrote some js so that when the user hits enter on the focused nav-item it expands.

I'm not sure if the js I wrote belongs in a new partials/javascripts, but I couldn't find a better place to put it. The problem is that when I build, the script is replaced by <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>. The temporary solution I used in my own project was just the stick the script at the bottom of the main.html override, and that works, but it doesn't seem like the right place to put it.

I think this is very close, so I would appreciate some more eyes on it. Thanks!

**Reply by squidfunk** [reply]
I'd be happy to improve the accessibility of the navigation. Could you create an issue for this, linking to your repository? This would be a good basis for discussion and taking action.

**Reply by dmundra** [reply]
A pull request would be great! The code changes look reasonable to me: https://github.com/squidfunk/mkdocs-material/compare/master...mnazzaro:mkdocs-material:master
